node/src/block_producer/vrf_evaluator/block_producer_vrf_evaluator_effects.rs

```diff
@@ -0,0 +1,131 @@
+    pub fn effects<S: Service>(self, _: &ActionMeta, store: &mut Store<S>) {
+        // TODO(adonagy): once block producer is enabled
+        // if let Some(config) = store.state().block_producer.config() {
+        //     store.dispatch(BlockProducerVrfEvaluatorUpdateProducerAndDelegatesAction {
+        //         current_epoch_ledger_hash: self.epoch_data.ledger.hash,
+        //         next_epoch_ledger_hash: self.next_epoch_data.ledger.hash,
+        //         producer: config.pub_key.to_string(),
+        //     });
+        // }
+
+        // let vrf_evaluator_state = store.state().block_producer.vrf_evaluator();
+        if let Some(vrf_evaluator_state) = store.state().block_producer.vrf_evaluator() {
+            store.dispatch(BlockProducerVrfEvaluatorUpdateProducerAndDelegatesAction {
+                current_epoch_ledger_hash: self.epoch_data.ledger.hash,
+                next_epoch_ledger_hash: self.next_epoch_data.ledger.hash,
+                producer: vrf_evaluator_state.producer_pub_key.to_string(),
+            });
+        }
+    }
+}
+
+impl BlockProducerVrfEvaluatorEvaluateVrfAction {
+    pub fn effects<S: Service>(self, _: &ActionMeta, store: &mut Store<S>) {
+        store.service.evaluate(self.vrf_input);
+    }
+}
+
+// impl BlockProducerVrfEvaluatorEvaluationPendingAction {
+//     pub fn effects<S: Service>(self, _: &ActionMeta, store: &mut Store<S>) {}
+// }
+
+impl BlockProducerVrfEvaluatorEvaluationSuccessAction {
+    pub fn effects<S: Service>(self, _: &ActionMeta, store: &mut Store<S>) {
+        let vrf_evaluator_state = store.state().block_producer.vrf_evaluator();
+
+        if let Some(vrf_evaluator_state) = vrf_evaluator_state {
+            let next_slot = vrf_evaluator_state.latest_evaluated_slot + 1;
+            // TODO(adonagy): Can we get this from somewhere?
+            const SLOTS_PER_EPOCH: u32 = 7140;
+            // determine the epoch of the slot
+            if let (Some(current_epoch), Some(current_epoch_data), Some(next_epoch_data)) = (
+                &vrf_evaluator_state.current_epoch,
+                &vrf_evaluator_state.current_epoch_data,
+                &vrf_evaluator_state.next_epoch_data,
+            ) {
+                let current_epoch_end = current_epoch * SLOTS_PER_EPOCH + SLOTS_PER_EPOCH - 1;
+                let next_epoch_end = (current_epoch + 1) * SLOTS_PER_EPOCH + SLOTS_PER_EPOCH - 1;
+
+                // slot is in the current epoch
+                if next_slot <= current_epoch_end {
+                    let vrf_input: VrfEvaluatorInput = VrfEvaluatorInput::new(
+                        current_epoch_data.seed.clone(),
+                        current_epoch_data.delegator_table.clone(),
+                        next_slot,
+                        current_epoch_data.total_currency,
+                        current_epoch_data.ledger.clone(),
+                    );
+                    store.dispatch(BlockProducerVrfEvaluatorEvaluateVrfAction { vrf_input });
```

Comment:
Enabling condition of `BlockProducerVrfEvaluatorEvaluateVrfAction` should contain all the above checks.

Reply:
One way to do that is to extract all this logic into function which takes state and returns next `VrfEvaluatorInput`. Then in the enabling condition you can verify that passed inputs match the ones returned from that function. For data in `Arc` we can compare just the pointers, otherwise comparing delegate table might be a bit expensive.